modules/compatibility-tests/src/integrationTest/java/org/apache/ignite/internal/PersistentCompatibilityTest.java

```diff
@@ -0,0 +1,218 @@
+
+            prepareTable(baseIgnite, TABLE_WITHOUT_DELTA_FILES);
+
+            // Checkpoint for TABLE_WITHOUT_DELTA_FILES and compact all delta files.
+            doCheckpoint(false);
+
+            prepareTable(baseIgnite, TABLE_WITH_DELTA_FILES);
+            prepareTable(baseIgnite, TABLE_WITH_NEW_DELTA_FILES);
+
+            // Checkpoint for TABLE_WITH_DELTA_FILES and TABLE_WITH_NEW_DELTA_FILES, cancels compaction to leave delta files.
+            doCheckpoint(true);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static void createAndPopulateTable(Ignite baseIgnite, String tableName) {
+        sql(baseIgnite, "CREATE TABLE " + tableName + " (id INT PRIMARY KEY, name VARCHAR)");
+        sql(baseIgnite, "INSERT INTO " + tableName + " (id, name) VALUES (1, 'test'), (2, 'deleted_row'), (3, 'original_value')");
+    }
+
+    private static void prepareTable(Ignite baseIgnite, String tableName) {
+        sql(baseIgnite, "DELETE FROM " + tableName + " WHERE id = 2");
+
+        Transaction tx = baseIgnite.transactions().begin();
+        baseIgnite.sql().execute(tx, "UPDATE " + tableName + " SET name = 'rolled_back_value' WHERE id = 3").close();
+        tx.rollback();
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {TABLE_WITH_DELTA_FILES, TABLE_WITHOUT_DELTA_FILES})
+    void testNewVersion(String tableName) throws IOException {
+        checkRows(tableName);
+    }
+
+    @Test
+    void testNewVersionWithNewDeltaFiles() throws IOException {
+        sql("INSERT INTO " + TABLE_WITH_NEW_DELTA_FILES + " (id, name) VALUES (4, 'new_row')");
+        sql("UPDATE " + TABLE_WITH_NEW_DELTA_FILES + " SET name = 'updated_value' WHERE id = 1");
+
+        doCheckpoint(true);
+
+        List<List<Object>> rows = sql("select * from " + TABLE_WITH_NEW_DELTA_FILES + " order by id");
+
+        assertThat(rows.size(), is(3));
+        assertThat(rows.get(0).get(1), is("updated_value"));
+        assertThat(rows.get(1).get(1), is("original_value"));
+        assertThat(rows.get(2).get(1), is("new_row"));
+    }
+
+    private void doCheckpoint(boolean cancelCompaction) throws IOException {
+        try (IgniteClient client = cluster.createClient()) {
+            JobDescriptor<Boolean, Void> job = JobDescriptor.builder(CheckpointJob.class)
+                    .units(new DeploymentUnit(CheckpointJob.class.getName(), "1.0.0")).build();
+
+            JobTarget jobTarget = JobTarget.anyNode(client.cluster().nodes());
+
+            client.compute().execute(jobTarget, job, cancelCompaction);
+        }
+    }
+
+    private <T, R> void deployCheckpointJob() throws IOException {
+        Path jarFile = createJar(CheckpointJob.class);
+
+        HttpResponse<Object> deploy = deploy(CheckpointJob.class.getName(), "1.0.0", jarFile.toFile());
+        assertThat(deploy.status(), is(HttpStatus.OK));
+    }
+
+    private Path createJar(Class<?> clazz) throws IOException {
```

Comment:
This creates a jar containing just 1 file. Here it works, but if it uses some custom code/library, another approach might be needed (take a look at `jobs.gradle`). Though I like your approach here. At your discretion.